source/adapters/level_zero/context.hpp

```diff
@@ -316,34 +331,77 @@ struct ur_context_handle_t_ : _ur_object {
     if (HostVisible) {
       if (Device) {
         auto EventCachesMap =
-            WithProfiling ? &EventCachesDeviceMap[0] : &EventCachesDeviceMap[1];
+            WithProfiling ? &EventCachesDeviceMap[HostVisibleProfilingCacheType]
+                          : &EventCachesDeviceMap[HostVisibleRegularCacheType];
+        if (EventCachesMap->find(Device) == EventCachesMap->end()) {
```

Comment:
These checks are for simply accessing event cache based on host-visible, profiling, and device. 
Host-visible has its own indexes and likewise for profiling. The only times we need to insert/emplace back is when we specify a device and that device is not already in the event cache map.

Reply:
but this requires to call find on slow container on every getEventCache call, that is not good for performance, can you preallocate all of this when you create a context? at that point you know how many devices it would contain so you can pre-populate all the maps , so here you just have always everything already available.

Reply:
That makes sense, added that in. Thanks!